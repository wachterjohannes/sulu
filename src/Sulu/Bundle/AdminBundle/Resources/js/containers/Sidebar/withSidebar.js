// @flow
import {autorun} from 'mobx';
import type {Component} from 'react';
import {buildHocDisplayName} from '../../services/react';
import type {SidebarConfig} from './types';
import sidebarStore from './stores/SidebarStore';

export default function withSidebar(
    Component: Class<Component<*, *>>,
    sidebar: () => SidebarConfig
) {
    const WithSidebarComponent = class extends Component {
        static hasSidebar = true;

        sidebarDisposer: Function;

        componentWillMount() {
            if (super.hasOwnProperty('sidebarDisposer')) {
                throw new Error('Component passed to withSidebar cannot declare a property called "sidebarDisposer".');
            }

            if (super.componentWillMount) {
                super.componentWillMount();
            }

            this.sidebarDisposer = autorun(() => {
                sidebarStore.setConfig(sidebar.call(this));
            });
        }

        componentWillUnmount() {
            if (super.componentWillUnmount) {
                super.componentWillUnmount();
            }

            this.sidebarDisposer();
        }
    };

    WithSidebarComponent.displayName = buildHocDisplayName('withSidebar', Component);

    return WithSidebarComponent;
}
