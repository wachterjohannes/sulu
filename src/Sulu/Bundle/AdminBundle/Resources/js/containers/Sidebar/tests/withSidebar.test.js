/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import {observable} from 'mobx';
import {mount, render} from 'enzyme';
import sidebarStore from '../stores/SidebarStore';
import withSidebar from '../withSidebar';

jest.mock('../stores/SidebarStore', () => ({
    setConfig: jest.fn(),
}));

test('Pass props to rendered component', () => {
    const Component = class Component extends React.Component {
        render() {
            return <h1>{this.props.title}</h1>;
        }
    };

    const ComponentWithSidebar = withSidebar(Component, () => {});

    expect(render(<ComponentWithSidebar title="Test" />)).toMatchSnapshot();
});

test('Bind sidebar method to component instance', () => {
    const Component = class Component extends React.Component {
        sidebarView = 'preview';

        render() {
            return <h1>Test</h1>;
        }
    };

    const ComponentWithSidebar = withSidebar(Component, function() {
        return {
            view: this.sidebarView,
        };
    });

    mount(<ComponentWithSidebar />);
    expect(sidebarStore.setConfig).toBeCalledWith({
        view: 'preview',
    });
});

test('Call life-cycle events of rendered component', () => {
    const Component = class Component extends React.Component {
        componentWillMount = jest.fn();
        componentWillUnmount = jest.fn();
        render = jest.fn();
    };

    const ComponentWithSidebar = withSidebar(Component, () => {});

    let component = mount(<ComponentWithSidebar />);
    expect(component.instance().componentWillMount).toBeCalled();
    expect(component.instance().render).toBeCalled();

    const componentWillUnmount = component.instance().componentWillUnmount;
    component.unmount();
    expect(componentWillUnmount).toBeCalled();
});

test('Recall sidebar-function when changing observable', () => {
    const Component = class Component extends React.Component {
        @observable sidebarView = 'preview';

        render() {
            return <h1>Test</h1>;
        }
    };

    const ComponentWithSidebar = withSidebar(Component, function() {
        return {view: this.sidebarView};
    });

    let component = mount(<ComponentWithSidebar />);

    expect(sidebarStore.setConfig).toBeCalledWith({
        view: 'preview',
    });

    component.instance().sidebarView = 'test';
    expect(sidebarStore.setConfig).toBeCalledWith({
        view: 'test',
    });
});

test('Throw error when component has property sidebarDisposer', () => {
    const Component = class Component extends React.Component {
        sidebarDisposer = true;

        render() {
            return <h1>Test</h1>;
        }
    };

    const ComponentWithSidebar = withSidebar(Component, function() {
        return {disableAll: this.test};
    });

    expect(() => mount(<ComponentWithSidebar />))
        .toThrowError('Component passed to withSidebar cannot declare a property called "sidebarDisposer".');
});
