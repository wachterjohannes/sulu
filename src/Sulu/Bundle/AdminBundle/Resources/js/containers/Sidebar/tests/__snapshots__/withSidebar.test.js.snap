// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pass props to rendered component 1`] = `
<h1>
  Test
</h1>
`;
